### PR TITLE
consolidated s390 device configuration

### PR DESCRIPTION
**Pull request by steffen-maier**
Consolidate the persistent and dynamic configuration of s390-specific devices in Linux distributions by delegating the configuration to the existing framework `zdev` from s390-tools.

This pull request completes consolidated s390 device configuration in `anaconda`.
This also fixes a newly discovered bug about missing persistent device configuration if device is enabled outside of anaconda, e.g. using rd.zfcp from dracut.
Some of the commits [see their descriptions] **depend** on certain commits from:
* https://github.com/ibm-s390-linux/s390-tools/pull/158 and thus https://github.com/ibm-s390-linux/s390-tools/releases/tag/v2.31.0.
* https://github.com/dracutdevs/dracut/pull/2534 (see also https://github.com/rhinstaller/anaconda/pull/5250#issuecomment-1887152134 and https://github.com/rhinstaller/anaconda/pull/5250#issuecomment-1887434454)

This PR works together with https://github.com/storaged-project/blivet/pull/1162.

Zdev's job is to perform low-level configuration after which the user gets architecture-independent objects such as block devices, SCSI devices, or network interfaces. Those can and should in turn be configured with existing common code mechanisms. So there's a clear separated layering for configuration duties.

In particular, the s390-specific devices currently are: DASD (traditional disk), ZFCP (scsi), and ZNET representing channel-attached network (QETH incl. OSA and HiperSockets, LCS, CTC). Zdev has a stable command line user interface and abstracts from sysfs and from a persistent configuration representation. Zdev encapsulates configuration details. Systems management code can simply delegate configuration to zdev and thus reduce architecture-specific code.

_This improves user experience, serviceability, maintainability, and reduces test effort._

@jstodola @poncovka @sharkcz

Even though this is a draft pull request, I would appreciate review comments.
It's only a draft until we sorted out the dependencies and merge order of pull requests for different related projects. Some thoughts in the integration process are in https://github.com/rhinstaller/anaconda/pull/5250#issuecomment-1887522104.

I developed and tested this locally with updates.img based on code (anaconda+blivet) branched closely enough for the updates to work with RHEL9.1 products.img (current when I started development).
This is a forward port, which applied pretty much automatically to the new base.

### CODE REVIEW

**Comment by jkonecny12** [comment]
Hello @steffen-maier, you are targeting this change into Fedora-39, however, Fedora 39 is in final freeze so until we have a bug with approved exception we can't get it in. I would suggest you to switch branch to `master` instead and focus on Fedora 40.

File: pyanaconda/modules/storage/dasd/discover.py
Comment (VladimirSlavik) [inline pyanaconda/modules/storage/dasd/discover.py:65]: I think this is the bit that breaks `DASDTasksTestCase.test_discovery`, but check yourself.

**Review by VladimirSlavik — COMMENTED** [review]
Thank you! I'll defer to @poncovka, @rvykydal and @vojtechtrefny for more details. I am not sure about the high level matters, such as if this should be all somewhere in blivet or not.

Apart from the failing tests, here's what I found...

**Review by jstodola — COMMENTED** [review]
Since `chzdev` and `/lib/s390-tools/zdev-to-rd.znet` are used, anaconda.spec should be updated so that `anaconda-core` requires the `s390utils-core` package (assuming the `zdev-to-rd.znet` helper script will be placed there) and also the minimal version of  `s390utils-core` should be specified once it is known.

File: pyanaconda/modules/storage/reset.py
Comment (poncovka) [inline pyanaconda/modules/storage/reset.py:82]: I could image to have all these calls in the `blivet` or `libblockdev` library and use their abstraction, but let's ask @vojtechtrefny first. Anyway, it is a very low-level code for Anaconda.

Comment (steffen-maier) [reply pyanaconda/modules/storage/reset.py:82]: > I could image to have all these calls in the `blivet` or `libblockdev` library and use their abstraction, but let's ask @vojtechtrefny first. Anyway, it is a very low-level code for Anaconda.

FWIW, this is by no means limited to handling block devices, so I'm not sure blivet or libblockdev would be a good home. The import is for persistent configuration of s390 devices of _any_ type, in particular including network.
I just used the hook here because we already had zfcp hooked here and this was promising for me and worked well during function test. I think we could even perform the import before anaconda starts, it just needs to be after switch root of the anaconda initrd, so we can import into the (volatile) root-fs of stage2 where anaconda runs in.

**Comment by steffen-maier** [comment]
> Since `chzdev` and `/lib/s390-tools/zdev-to-rd.znet` are used, anaconda.spec should be updated so that `anaconda-core` requires the `s390utils-core` package (assuming the `zdev-to-rd.znet` helper script will be placed there) and also the minimal version of `s390utils-core` should be specified once it is known.

I have already prepared such change commit but defer pushing until I looked at the unit tests fails, so I don't trigger the git workflows too often and unnecessarily.
Until then, I would like to understrand the background. IIRC, the "test" mode of anaconda, where you could run it as a process in a regular running Linux instance (as opposed to the dedicated anaconda initrd&stage2 environment), was removed long ago. For such use case, I would have immediately understood the reason for rpm package dependencies. Setting aside anaconda development using -devel sub packages, I wonder if any user would install an anaconda "binary" rpm and thus need the package dependencies.
Do we add the requirements because this is basic input for lorax (if that is still used)? AFAIK, lorax introduced the approach of packaging full rpm content into an installer image, and then have lorax config to prune certain files from the image. I.e. an exclude list approach instead of an allow list approach as what was used before lorax (anaconda's former build scripts incl. scripts/mk-images used for composes ?).

File: pyanaconda/modules/network/nm_client.py
Comment (steffen-maier) [reply pyanaconda/modules/network/nm_client.py:1378]: I think it makes no sense anymore to actually call `_get_dracut_znet_argument_from_connection()` because all it does is delegate to the external helper program called via `util.execWithCapture()`. For the sake of completeness, I added `get_interface_name` to the mock objects even though it is unused. 
Let's see if this works as I hope: https://github.com/rhinstaller/anaconda/pull/5250/commits/80c3a42087655d9700417eb1d5c7983310b3a5ae

File: pyanaconda/modules/storage/dasd/discover.py
Comment (steffen-maier) [reply pyanaconda/modules/storage/dasd/discover.py:65]: Hopefully this fixup of the unit test works: https://github.com/rhinstaller/anaconda/pull/5250/commits/3168bff6c90c4dcb6072d694659a2964fdc9491d

**Comment by steffen-maier** [comment]
> Hello @steffen-maier! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:
> 
> * In the file [`pyanaconda/modules/network/nm_client.py`](https://github.com/rhinstaller/anaconda/blob/3168bff6c90c4dcb6072d694659a2964fdc9491d/pyanaconda/modules/network/nm_client.py):
> 
> > [Line 38:1](https://github.com/rhinstaller/anaconda/blob/3168bff6c90c4dcb6072d694659a2964fdc9491d/pyanaconda/modules/network/nm_client.py#L38): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> 
> ##### Comment last updated at 2023-10-18 17:30:34 UTC

Unfortunately, I'm not sure why that is. I fail to see how my code change introduced that. The only preceding non-import statement I can see is:
https://github.com/rhinstaller/anaconda/blob/3168bff6c90c4dcb6072d694659a2964fdc9491d/pyanaconda/modules/network/nm_client.py#L22
But that was already there before my changeset.

**Comment by VladimirSlavik** [comment]
> [Line 38:1](https://github.com/rhinstaller/anaconda/blob/3168bff6c90c4dcb6072d694659a2964fdc9491d/pyanaconda/modules/network/nm_client.py#L38): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file

Feel free to ignore these if not applicable. The GI imports always trip it up.

**Comment by VladimirSlavik** [comment]
> lorax

Yes, that is still used. However, for the long term we are trying to move towards having the dependencies in anaconda metapackages `anaconda-install-{env,img}-deps` and minimize the lorax part. If there is no technical reason to have the new dependency in lorax, such as removing large amounts of unused data from the installed package, then please add is as a dependency here.

By the way, I can't find any package with these tools?

**Comment by steffen-maier** [comment]
> By the way, I can't find any package with these tools?

You mean the helper scripts `/lib/s390-tools/zdev-...` being called by anaconda and blivet in this PR? If so:

> Some of the commits [see their descriptions] depend on certain commits from https://github.com/ibm-s390-linux/s390-tools/pull/158.

As Jan stated earlier, they will end up in a **future** release of `s390utils-core`.

**Comment by VladimirSlavik** [comment]
Actually, I can't find even the package `s390utils-core`. Or is that supposed to not exist yet? The commits you reference don't change any spec file...

**Comment by steffen-maier** [comment]
> Actually, I can't find even the package `s390utils-core`. Or is that supposed to not exist yet? The commits you reference don't change any spec file...

Good point. I pointed to the upstream repo, which does not have any packaging definitions. Downstream packaging is by means of https://src.fedoraproject.org/rpms/s390utils/blob/rawhide/f/s390utils.spec#_206 already creating s390utils-core. Maybe I should add spec file changes for my upstream changes to my https://github.com/steffen-maier/s390utils/pull/1, which currently only exists of pure downstream contributions.

**Comment by steffen-maier** [comment]
How can I easily (without containers or other non-trivial setup) run the unit tests locally? Pointers to doc might be sufficient.
I messed it up again and it took very long for the git workflow to run after I had pushed updates, so I'm looking for more efficient turnaround times.

**Comment by sharkcz** [comment]
> Actually, I can't find even the package `s390utils-core`. Or is that supposed to not exist yet? The commits you reference don't change any spec file...

it is a s390x specific package, not available anywhere else (x86_64, ...)

**Comment by VladimirSlavik** [comment]
Generally, `make -f Makefile.am container-ci` should be sufficient locally - it will grab the missing container and run unit tests.

**Comment by VladimirSlavik** [comment]
As for https://github.com/storaged-project/blivet/pull/1162#issuecomment-1771443131, the docs have a section about [patching the container used for unit tests](https://anaconda-installer.readthedocs.io/en/latest/testing.html#run-unit-tests-with-patched-pykickstart-or-other-libraries), which might help.

**Comment by steffen-maier** [comment]
> * Generally, `make -f Makefile.am container-ci` should be sufficient locally - it should grab the missing container and run unit tests. I'd give it 90% confidence that it will "just work", unless you're not on x86_64.

FYI: I had to manually load the ext4 kernel module on my local container host `sudo modprobe ext4`, otherwise a number of additional unit tests failed (that worked here in the git workflow) because ext4 was not available and it could not load the module either (non-root and in container?).

**Comment by steffen-maier** [comment]
I'm still working on the unit tests, got the running locally but then got distracted. Maybe I will need help with the unit tests.

About the current re-run of the git workflow from https://github.com/rhinstaller/anaconda/actions/runs/6577747989/job/18665300996?pr=5250:
These are still caused by my code changes here:
`=========================== short test summary info ============================`
`FAILED unit_tests/pyanaconda_tests/modules/network/test_module_network_nm_client.py::NMClientTestCase::test_get_dracut_arguments_from_connection`
`FAILED unit_tests/pyanaconda_tests/modules/storage/test_module_dasd.py::DASDTasksTestCase::test_discovery
FAILED`

The following ones are new and I don't see how my unchanged code could have cause these. Maybe there is a new blivet version being pulled in and the anaconda unit tests were not yet updated for those blivet changes unrelted to this PR?:
`unit_tests/pyanaconda_tests/modules/storage/partitioning/test_module_part_blivet.py::BlivetPartitioningInterfaceTestCase::test_request_handler`
`FAILED unit_tests/pyanaconda_tests/modules/storage/partitioning/test_module_part_blivet.py::BlivetPartitioningInterfaceTestCase::test_send_request`
`FAILED unit_tests/pyanaconda_tests/modules/storage/partitioning/test_module_part_blivet.py::BlivetPartitioningInterfaceTestCase::test_storage_handler`

> from blivet.tasks.fslabeling import Ext2FSLabeling, FATFSLabeling, JFSLabeling, ReiserFSLabeling, XFSLabeling, NTFSLabeling
> E   ImportError: cannot import name 'JFSLabeling' from 'blivet.tasks.fslabeling' (/usr/lib/python3.12/site-packages/blivet/tasks/fslabeling.py)

> raise UnsupportedPartitioningError("Missing support for Blivet-GUI") from e
> E  pyanaconda.modules.common.errors.storage.UnsupportedPartitioningError: Missing support for Blivet-GUI

`FAILED unit_tests/pyanaconda_tests/ui/test_simple_ui.py::SimpleUITestCase::test_gui`

> `blivet-gui-partitioning` vs. `interactive-partitioning`

`====== 6 failed, 2031 passed, 1 xfailed, 21 warnings in 238.13s (0:03:58) ======`

**Comment by steffen-maier** [comment]
> I'm still working on the unit tests, got the running locally but then got distracted. Maybe I will need help with the unit tests.
> 
> About the current re-run of the git workflow from https://github.com/rhinstaller/anaconda/actions/runs/6577747989/job/18665300996?pr=5250:
> These are still caused by my code changes here:
> `=========================== short test summary info ============================`
> `FAILED unit_tests/pyanaconda_tests/modules/network/test_module_network_nm_client.py::NMClientTestCase::test_get_dracut_arguments_from_connection`

fixed

> `FAILED unit_tests/pyanaconda_tests/modules/storage/test_module_dasd.py::DASDTasksTestCase::test_discovery FAILED`

still working on that, so no need to trigger a re-run of the unit tests

File: pyanaconda/modules/storage/reset.py
Comment (rvykydal) [reply pyanaconda/modules/storage/reset.py:82]: @steffen-maier what exactly does the import do (applying the data)? Is it safe (sane) to call it multiple times?
The storage rescan can happen quite a few times during interactive installation (and the iscsi/fcoe/zfcp.startup() would do the startup only during the first call) so I think we should try to find a better place. Also given the import regards also network devices.

Comment (steffen-maier) [reply pyanaconda/modules/storage/reset.py:82]: https://github.com/ibm-s390-linux/s390-tools/pull/158/commits/7c21e53376238412ebb0977b5fec2cab8d2606fe#diff-c2758979d9762cd0f733650085a42de8533819b87c19665dc42f0c44e6442561 ("zdev/dracut: retain early persistent config over switch root") exports the persistent s390 device configuration established during early initrd (parsing result of rd.dasd, rd.zfcp, rd.znet) to `/run/zdev.initrd.config`, which is in a tmpfs that remains visible with its content even after dracut's switch-root. It's similar to and a generalization of anaconda-copy-s390ccwconf.sh (which this PR removes https://github.com/rhinstaller/anaconda/pull/5250/files/61aff0c3cc1fbedd142d869a9f8f106e723efdee#diff-f4a4c6c6e09f6db50c1840d9e254e2719ba33adae3878e7c9a1a99ed58e21d5c).

Here when we import the persistent config, we are already in the anaconda stage2 root-fs. Before the (first) import it has no s390 device persistent configuration. The import brings the persistent config into the stage2 environment.
An import **merges** with any persistent config that might already exist in the (stage2) environment. Hence, it should be safe to import multiple times (even if unnecessary).

Later during stage2, the user/kickstart can establish additional persistent config. Because other anaconda code (such as ZFCPDiskDevice.dracut_setup_args()) also needs to read any imported or established persistent config, the import needs to happen early enough.

Finally, the set union of persistent config gets persisted into the installation target (https://github.com/rhinstaller/anaconda/pull/5250/commits/022465897dac2d485c54cb019162642277df3d23#diff-d1f055451308b33c2271dfe2958a633fadc7a53acb9c5e7019ad1c45fc3ff818R293 ("write persistent config of any (dasd,zfcp,znet) s390 devices to sysrroot (#1802482,#1937049)")).) That kind of replaces the former different `...write...` methods for the different s390 device types.

I need help regarding a better code location for the import as that's way beyond my understanding of anaconda.

**Comment by steffen-maier** [comment]
FYI, regarding a new failing check: **Anaconda Web UI / trigger (pull_request_target) Failing after 11m**
https://github.com/rhinstaller/anaconda/actions/runs/6867742140/job/18676757199?pr=5250
Looks like I cannot do anything about it (and I doubt the backend code changes here would influence/conflict with the new WebUI).

```
Trigger anaconda run
Run git clone --depth=[1](https://github.com/rhinstaller/anaconda/actions/runs/6867742140/job/18676757199?pr=5250#step:5:1) https://github.com/cockpit-project/bots
  git clone --depth=1 https://github.com/cockpit-project/bots
  mkdir -p ~/.config/cockpit-dev
  echo '' > ~/.config/cockpit-dev/github-token
  bots/tests-trigger --repo rhinstaller/anaconda 5[2](https://github.com/rhinstaller/anaconda/actions/runs/6867742140/job/18676757199?pr=5250#step:5:2)[5](https://github.com/rhinstaller/anaconda/actions/runs/6867742140/job/18676757199?pr=5250#step:5:5)0 fedora-rawhide-boot/anaconda-pr-5250@rhinstaller/anaconda-webui
  shell: sh -e {0}
Cloning into 'bots'...
Pull request author 'steffen-maier' isn't allowed. Override with --allow. <=====
Error: Process completed with exit code 1.
```

**Comment by rvykydal** [comment]
> The following ones are new and I don't see how my unchanged code could have cause these. Maybe there is a new blivet version being pulled in and the anaconda unit tests were not yet updated for those blivet changes unrelted to this PR?: `unit_tests/pyanaconda_tests/modules/storage/partitioning/test_module_part_blivet.py::BlivetPartitioningInterfaceTestCase::test_request_handler` `FAILED unit_tests/pyanaconda_tests/modules/storage/partitioning/test_module_part_blivet.py::BlivetPartitioningInterfaceTestCase::test_send_request` `FAILED unit_tests/pyanaconda_tests/modules/storage/partitioning/test_module_part_blivet.py::BlivetPartitioningInterfaceTestCase::test_storage_handler`
> 
> > from blivet.tasks.fslabeling import Ext2FSLabeling, FATFSLabeling, JFSLabeling, ReiserFSLabeling, XFSLabeling, NTFSLabeling
> > E   ImportError: cannot import name 'JFSLabeling' from 'blivet.tasks.fslabeling' (/usr/lib/python3.12/site-packages/blivet/tasks/fslabeling.py)

Yes, you can ignore them if possible.

File: pyanaconda/modules/storage/reset.py
Comment (rvykydal) [reply pyanaconda/modules/storage/reset.py:82]: Thank you for explanation, I was afraid the import does some activation or setup beyond the persistent config update. This way it seems safe and fine at this point. I will try to think about some better place.
And thank you for updating the tests!

**Comment by steffen-maier** [comment]
> > `FAILED unit_tests/pyanaconda_tests/modules/storage/test_module_dasd.py::DASDTasksTestCase::test_discovery FAILED`
> 
> still working on that, so no need to trigger a re-run of the unit tests

fixed

**Comment by steffen-maier** [comment]
Jenkins, ok to test.

**Comment by steffen-maier** [comment]
> Some of the tests here run only after they're manually confirmed.

I'm not sure the unit-tests were already triggered since I had force pushed the finally fixed and squashed code one day ago (https://github.com/rhinstaller/anaconda/pull/5250#issuecomment-1812576596). The current state I can see here is:
`unit-tests Expected — Waiting for status to be reported`
I would like to confirm that the unit tests related to the code change here are OK.
Could someone please trigger the unit tests unless they have already started? @rvykydal  @VladimirSlavik 
Once I could confirm, I'm going to add one more unit test update commit (https://github.com/steffen-maier/anaconda/commit/9873c62f8e1a0ff0dc2e713dc00e64f4ac6636c5) to anticipate the related blivet changes from https://github.com/storaged-project/blivet/pull/1162#issuecomment-1771443131; that way the code would be final over both packages.

**Comment by steffen-maier** [comment]
tests good
https://github.com/rhinstaller/anaconda/actions/runs/6878142943/job/18923478959?pr=5250#step:6:2868
```
PASS: cppcheck/runcppcheck.sh
PASS: gettext_tests/click.py
PASS: gettext_tests/style_guide.py
PASS: gettext_tests/contexts.py
PASS: gettext_tests/canary_tests.sh
PASS: glade_tests/glade_tests.sh
PASS: shellcheck/run_shellcheck.sh
PASS: ruff/run_ruff.sh
PASS: unit_tests/unit_tests.sh
============================================================================
Testsuite summary for anaconda 40.11
============================================================================
# TOTAL: 9
# PASS:  9
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

File: pyanaconda/modules/storage/reset.py
Comment (rvykydal) [reply pyanaconda/modules/storage/reset.py:82]: > Thank you for explanation, I was afraid the import does some activation or setup beyond the persistent config update. This way it seems safe and fine at this point. I will try to think about some better place. And thank you for updating the tests!

I wonder if it would make sense to import the state from initramfs by a systemd service (the services we use now: https://github.com/rhinstaller/anaconda/tree/master/data/systemd)

Comment (steffen-maier) [reply pyanaconda/modules/storage/reset.py]: https://github.com/rhinstaller/anaconda/pull/5250/commits/e94552e287c74cde37e330674ca93c38bd942d9b uses a systemd service and worked well

Comment (rvykydal) [reply pyanaconda/modules/storage/reset.py]: Nice, thank you!

**Comment by rvykydal** [comment]
@steffen-maier is this update in Anaconda (and perhaps related python-blivet changes in https://github.com/storaged-project/blivet/pull/1162) required to not block the whole effort (changes in the dracut PR https://github.com/dracutdevs/dracut/pull/2534) and/or in https://github.com/steffen-maier/s390utils/pull/1) or is there a chance that the old approach will keep working for some time before applying the Anaconda (blivet) changes?

I think the Anaconda and blivet PRs are ready and I'd prefer the switch to zdev solution, and I'll approve anaconda PR when we make sure test would pass and perhaps update the FIXME in the spec file, but it is possible that we might face resource issues for applying the changes, so I'd like to clear up what would be the risks there.

**Comment by steffen-maier** [comment]
Updated the commit descriptions with the actual upstream commits since https://github.com/ibm-s390-linux/s390-tools/pull/158 got merged.
Clarified the explicit package dependency in spec file, to model the dependencies more fine granular: commit  ("DASDDiscoverTask...") does only depend on s390utils-core but not a particular new version, only commit ("network...") depends on a (future) new version of s390utils-core (originated in  https://github.com/rhinstaller/anaconda/pull/5250#pullrequestreview-1682480291).
**Added a similar version dependency on dracut-network in the anaconda spec file, because commit ("network: use consolidated s390 device configuration (#1802482,#1937049)") can only get the rd.znet boot persistent device config from initrd (and transfer it to the installed system so that'll boot with a network connection) with an updated dracut module "znet" from https://github.com/dracutdevs/dracut/pull/2534.**
Squashed the fixup commit moving the import of initrd zdev config from anaconda python code to a new anaconda systemd service (https://github.com/rhinstaller/anaconda/pull/5250#discussion_r1409695178).
No code changes.

**Comment by steffen-maier** [comment]
Rebased to latest master and force pushed to resolve the merge conflict in [pyanaconda/modules/storage/installation.py](https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/modules/storage/installation.py) due to commit 9b826817d281 ("Add a simple NVMe module for NVMe Fabrics support").

Also fixed:
> Clarified the explicit package dependency in spec file, to model the dependencies more fine granular: commit ("DASDDiscoverTask...") does only depend on s390utils-core but not a particular new version, only commit ("network...") depends on a (future) new version of s390utils-core (originated in [#5250 (review)](https://github.com/rhinstaller/anaconda/pull/5250#pullrequestreview-1682480291)).

The new s390utils-core version is already required one commit earlier by ("write persistent config..."), so I moved the version dependency there.

**Comment by steffen-maier** [comment]
> @steffen-maier is this update in Anaconda (and perhaps related python-blivet changes in [storaged-project/blivet#1162](https://github.com/storaged-project/blivet/pull/1162)) required to not block the whole effort (changes in the dracut PR [dracutdevs/dracut#2534](https://github.com/dracutdevs/dracut/pull/2534)) and/or in [steffen-maier/s390utils#1](https://github.com/steffen-maier/s390utils/pull/1)) 

@rvykydal this anaconda PR and the connected https://github.com/storaged-project/blivet/pull/1162 are vital core parts of the consolidation endeavor.
However, they do not block the dracut PR as that one is independent. Actually, this anaconda PR here has one commit that depends on the dracut PR (https://github.com/rhinstaller/anaconda/pull/5250#issuecomment-1887152134) .
As much as I would like to integrate anaconda&blivet soon, I also would like to avoid breaking installability of Fedora Rawhide s390x for an indetermined amount of time (until the dracut changes hopefully land upstream and the Fedora dracut packaging rebases onto that).

> or is there a chance that the old approach will keep working for some time before applying the Anaconda (blivet) changes?

Yes, as long as we hold back https://github.com/steffen-maier/s390utils/pull/1 the old mechanism is still there (@sharkcz please do _not_ pull that one for the time being :-) ).
However, we should make sure that the s390utils downstream PR gets integrated tightly after anaconda&blivet(&dracut) to land together in a distro release. Otherwise, it could contradict the consolidation effort: Dracut(+s390-tools) currently have 3 different methods to configure s390 devices during early initrd, Fedora has its own custom downstream method to configure s390 devices after initrd switch root; if we add another new method without removing the old one it would seem the opposite of consolidation to me -- plus it might cause functional trouble due to concurrent double configuration of the same devices by different methods.

> I think the Anaconda and blivet PRs are ready and I'd prefer the switch to zdev solution, and I'll approve anaconda PR when we make sure test would pass and perhaps update the FIXME in the spec file,

Even though we would like to have all s390 device types use either the old mechanism or the zdev mechanism, I could factor the network commit depending the pending dracut PR into a new separate anaconda PR. This way, the existing anaconda&blivet PRs do not depend on dracut changes anymore and we could integrate them to reduce backlog. However, I'm a bit nervous about the dracut changes taking longer and we could end up with a distro using a different old mechanism just for s390 network device configuration.
Any idea how to help speed the required dracut review? @dtardon was the only one [providing review comments](https://github.com/dracutdevs/dracut/pull/2534#issuecomment-1805189234) so far and they were good ones, thank you!

@rvykydal, Please let me know if you would like to go the approach of factoring out the network part.

Note, however, that we still need a new version of s390utils(-core) including the content of https://github.com/ibm-s390-linux/s390-tools/pull/158 as prereq for anaconda&blivet.
@sharkcz, that's 24 commits that are upstream but there is no official upstream release on https://github.com/ibm-s390-linux/s390-tools/releases yet including those. Do you need such release to bring the content into a new rawhide build of s390utils or is there a different possibility (downstream patches for the time being)?

> but it is possible that we might face resource issues for applying the changes, so I'd like to clear up what would be the risks there.

It's not clear to me what efforts the integration would cause. I tried to provide a turn key solution. Maybe we can discuss and resolve offline how to minimize integration effort.

**Comment by sharkcz** [comment]
yes, I am going to look/review ASAP

**Comment by rvykydal** [comment]
@steffen-maier thank you for the details on the process, I still believe that when the dracut part is ready we will be able to switch anaconda.

**Comment by steffen-maier** [comment]
> Since `chzdev` and `/lib/s390-tools/zdev-to-rd.znet` are used, anaconda.spec should be updated so that `anaconda-core` requires the `s390utils-core` package (assuming the `zdev-to-rd.znet` helper script will be placed there) and also the minimal version of `s390utils-core` should be specified once it is known.

Now the upstream version v2.31.0 is known and I updated the commit touching the spec file ("write persistent config of any (dasd,zfcp,znet) s390 devices to sysroot (#1802482,#1937049)").

**Comment by steffen-maier** [comment]
This is not stale. It's just blocked by https://github.com/dracutdevs/dracut/pull/2534.

**Comment by jstodola** [comment]
Tested these changes against a recent RHEL-10 build and the changes work there fine.

**Comment by jstodola** [comment]
@steffen-maier is this still a draft?

The dracut changes have been merged in https://github.com/dracut-ng/dracut-ng/pull/73 and the changes are included in dracut-102-3.fc41, which is present in Fedora-41 composes.

Blivet changes are tracked in https://github.com/storaged-project/blivet/pull/1259 (the PR is also marked as Draft).

On RHEL-10 we merged the anaconda part first and then the blivet part and we have not noticed any problem, so I believe this PR is not blocked.

Since this feature has already been merged into RHEL-10 (https://github.com/rhinstaller/anaconda/pull/5676), removing the "port to RHEL10" label, as well as the "stale" and "blocked" labels, since I believe all prerequisites have been satisfied.

**Comment by steffen-maier** [comment]
Hi Jan, thanks for your updates. I also had the impression that dependencies are meanwhile fulfilled and I need to update this PR (such as dracut version in code [thanks for looking that up!] and remove the draft state), but I currently don't get to it.
(Wondering if that would also mean to remove the draft state from the blivet PR.)

**Comment by steffen-maier** [comment]
> The dracut changes have been merged in [dracut-ng/dracut-ng#73](https://github.com/dracut-ng/dracut-ng/pull/73) and the changes are included in dracut-102-3.fc41, which is present in Fedora-41 composes.

I updated the code to depend on dracut 102-3.
Originally thought 102-1 would be closer because the required code changes appeared upstream with dracut-ng 102 and Fedora dracut 102-1 rebased onto that:
> * Fri Jun 28 2024 Pavel Valena <pvalena@redhat.com> - 102-1
> - Update to dracut 102.

(https://koji.fedoraproject.org/koji/buildinfo?buildID=2490409)
but I also saw a 102-1 for Fedora-40 and wasn't sure if this could become a problem if 102-1 would be the minimum dependency
(https://koji.fedoraproject.org/koji/packageinfo?packageID=8714)
so I took 102-3 like you said.

I also removed the draft state so this can get merged.

**Comment by rvykydal** [comment]
Validation tests failure is expected, depends on the related blivet update.

**Review by rvykydal — APPROVED** [review]
Looks good to me.

**Comment by jstodola** [comment]
Quickly retested the changes using an updates.img provided by Radek. Tested against rawhide on a KVM and z/VM system, no issue found.

**Comment by rvykydal** [comment]
@steffen-maier it seems that we need to update the PR to current master so that `rpm-build:fedora-enl-x86_64`, `rpm-build:fedora-rawhide-x86_64` and `testing-farm:fedora-rawhide-x86_64` can be run successfuly (They pass in testing PR https://github.com/rhinstaller/anaconda/pull/5927).
Could you plaese rebase it? (Or I can rebase it for you if you don't mind.)

**Pull request merged**
